File: app/Ligare/Core/Classify.hs
module Ligare.Core.Classify where

import Ligare.Checker.Context (Context, lookupCtx)
import Ligare.Core.Syntax

classify :: Context -> Term -> Maybe Universe
classify _ (LitInt _) = Just UData
classify _ (LitBool _) = Just UData
classify _ (Lam _) = Just UData
classify ctx (App f _) = classify ctx f -- 应用程序的宇宙取决于函数
classify _ (PrimOp _) = Just UData
classify _ (Universe u) = Just u
classify _ AutoProof = Just UProof
classify _ RefParam = Just UData -- 精化参数是一个值
classify _ (Func {}) = Just UData -- 展开后是 lambda
classify ctx (Var i) = do
  t <- lookupCtx i ctx
  classify ctx t
classify ctx (Annot t _) = classify ctx t
classify ctx (ByProof t _) = classify ctx t
classify ctx (Let _ _ body _) = classify ctx body
classify ctx (IfThenElse _ t _) = classify ctx t
classify ctx (ProofBlock t) = classify ctx t -- proof block 的宇宙取决于其内容
classify _ (Pi _ _ _) = Just UProp
classify _ (Refine _ _ _) = Just UProp
classify _ (Builtin "int") = Just UProp -- int 优先作为约束
classify _ (Builtin "bool") = Just UProp
classify _ (Builtin "data") = Just UProp
classify _ (Builtin "theorem") = Just UTheorem
classify _ (Builtin "proof") = Just UProof
classify _ (Builtin "and") = Just UProp
classify _ (Builtin "or") = Just UProp
classify _ (Builtin "not") = Just UProp
classify _ (Builtin "implies") = Just UProp
classify _ (Builtin _) = Nothing -- 未知 builtin

File: src/core/classify.rs
use crate::checker::builtin::classify_builtin;
use crate::checker::context::Context;
use crate::core::syntax::{Term, Universe};

pub struct Classifier;

impl Classifier {
    pub fn classify(ctx: &Context<'_>, t: &Term<'_>) -> Option<Universe> {
        match t {
            Term::LitInt(_)
            | Term::LitBool(_)
            | Term::LitStr(_)
            | Term::Lam(_)
            | Term::NamedLam(_, _)
            | Term::PrimOp(_)
            | Term::RefParam => Some(Universe::UData),
            Term::App(f, _) => Self::classify(ctx, f),
            Term::Universe(u) => Some(*u),
            Term::AutoProof => Some(Universe::UProof),
            Term::Pi(_, _, _) | Term::Refine(_, _, _) => Some(Universe::UProp),
            // Variables in an empty or unknown context are conservatively
            // assumed to be data-relevant (this handles function parameters
            // during erasure, where the typing context is not available).
            Term::Var(i) => ctx
                .lookup(*i)
                .and_then(|ty| Self::classify(ctx, ty))
                .or(Some(Universe::UData)),
            Term::Annot(t, _) => Self::classify(ctx, t),
            Term::ByProof(Some(t), _) => Self::classify(ctx, t),
            Term::ByProof(None, _) => Some(Universe::UProof),
            Term::Let(_, _, body, _) => Self::classify(ctx, body),
            Term::IfThenElse(_, t, _) => Self::classify(ctx, t),
            Term::Builtin(name) | Term::Named(name) => {
                classify_builtin(name).or(Some(Universe::UData))
            }
            Term::UnionDef(..) => Some(Universe::UProp),
            Term::Variant(..) => Some(Universe::UData),
            Term::StructDef(..) => Some(Universe::UProp),
            Term::StructCons(..) => Some(Universe::UData),
            Term::StructProj(subject, _) => Self::classify(ctx, subject),
            Term::Match(_, branches) => {
                // Match type = type of first branch (all branches must agree)
                branches
                    .first()
                    .map(|(_, _, body)| Self::classify(ctx, body))
                    .flatten()
            }
        }
    }
}

pub fn classify(ctx: &Context<'_>, t: &Term<'_>) -> Option<Universe> {
    Classifier::classify(ctx, t)
}
